stats/barseries.cpp

```diff
@@ -0,0 +1,422 @@
+	valueVariable(valueVariable), valueBinNames(std::move(valueBinNames))
+{
+}
+
+BarSeries::BarSeries(QtCharts::QChart *chart, StatsAxis *xAxis, StatsAxis *yAxis,
+		     bool horizontal, const QString &categoryName,
+		     const std::vector<CountItem> &items) :
+	BarSeries(chart, xAxis, yAxis, horizontal, false, categoryName, nullptr, std::vector<QString>())
+{
+	for (const CountItem &item: items) {
+		StatsOperationResults res;
+		res.count = item.count;
+		double value = item.count;
+		add_item(item.lowerBound, item.upperBound, makeSubItems(value, item.label),
+			 item.binName, res, item.total, horizontal, stacked);
+	}
+}
+
+BarSeries::BarSeries(QtCharts::QChart *chart, StatsAxis *xAxis, StatsAxis *yAxis,
+		     bool horizontal, const QString &categoryName, const StatsVariable *valueVariable,
+		     const std::vector<ValueItem> &items) :
+	BarSeries(chart, xAxis, yAxis, horizontal, false, categoryName, valueVariable, std::vector<QString>())
+{
+	for (const ValueItem &item: items) {
+		add_item(item.lowerBound, item.upperBound, makeSubItems(item.value, item.label),
+			 item.binName, item.res, -1, horizontal, stacked);
+	}
+}
+
+BarSeries::BarSeries(QtCharts::QChart *chart, StatsAxis *xAxis, StatsAxis *yAxis,
+		     bool horizontal, bool stacked, const QString &categoryName, const StatsVariable *valueVariable,
+		     std::vector<QString> valueBinNames,
+		     const std::vector<MultiItem> &items) :
+	BarSeries(chart, xAxis, yAxis, horizontal, stacked, categoryName, valueVariable, std::move(valueBinNames))
+{
+	for (const MultiItem &item: items) {
+		StatsOperationResults res;
+		std::vector<std::pair<double, std::vector<QString>>> valuesLabels;
+		valuesLabels.reserve(item.countLabels.size());
+		int total = 0;
+		for (auto &[count, label]: item.countLabels) {
+			valuesLabels.push_back({ static_cast<double>(count), std::move(label) });
+			total += count;
+		}
+		add_item(item.lowerBound, item.upperBound, makeSubItems(valuesLabels),
+			 item.binName, res, total, horizontal, stacked);
+	}
+}
+
+BarSeries::~BarSeries()
+{
+}
+
+BarSeries::BarLabel::BarLabel(QtCharts::QChart *chart, const std::vector<QString> &labels, int bin_nr, int binCount) :
+	totalWidth(0.0), totalHeight(0.0), isOutside(false)
+{
+	items.reserve(labels.size());
+	for (const QString &label: labels) {
+		items.emplace_back(new QGraphicsSimpleTextItem(chart));
+		items.back()->setText(label);
+		items.back()->setZValue(ZValues::seriesLabels);
+		QRectF rect = items.back()->boundingRect();
+		if (rect.width() > totalWidth)
+			totalWidth = rect.width();
+		totalHeight += rect.height();
+	}
+	highlight(false, bin_nr, binCount);
+}
+
+void BarSeries::BarLabel::setVisible(bool visible)
+{
+	for (auto &item: items)
+		item->setVisible(visible);
+}
+
+void BarSeries::BarLabel::highlight(bool highlight, int bin_nr, int binCount)
+{
+	QBrush brush(highlight || isOutside ? darkLabelColor : labelColor(bin_nr, binCount));
```

Comment:
this whole `isOutside` thing is a good catch... I will need to go through all of the uses of `darkLabelColor` once this is in and deal with that for a dark theme...